repo2docker/buildpacks/pipfile/__init__.py

```diff
@@ -0,0 +1,133 @@
+                "pattern": r"\s*python-([0-9a-z\.]*)\s*",  # 'python-3.6'
+            },
+        ]
+
+        py_version = None
+        for file in files_to_search_in_order:
+            try:
+                with open(file["path"]) as f:
+                    for line in f:
+                        match = re.match(file["pattern"], line)
+                        if not match:
+                            continue
+                        py_version = match.group(1)
+                        break
+            except FileNotFoundError:
+                pass
+            if py_version:
+                break
+
+        # extract major.minor
+        if py_version:
+            if len(py_version) == 1:
+                self._python_version = self.major_pythons.get(py_version[0])
+            else:
+                # return major.minor
+                self._python_version = ".".join(py_version.split(".")[:2])
+            return self._python_version
+        else:
+            # use the default Python
+            self._python_version = self.major_pythons["3"]
+            return self._python_version
+
+    def get_assemble_scripts(self):
+        """Return series of build-steps specific to this repository.
+        """
+        # If we have a runtime.txt & that's set to python-2.7,
+        # requirements.txt will be installed in the *kernel* env
+        # and requirements3.txt (if it exists)
+        # will be installed in the python 3 notebook server env.
+        assemble_scripts = super().get_assemble_scripts()
+
+        if self.py2:
+            # using python 2 kernel,
+            # requirements3.txt allows installation in the notebook server env
+            nb_requirements_file = self.binder_path("requirements3.txt")
+            if os.path.exists(nb_requirements_file):
+                assemble_scripts.append(
+                    (
+                        "${NB_USER}",
+                        # want the $NB_PYHTON_PREFIX environment variable, not for
+                        # Python's string formatting to try and replace this
+                        '${{NB_PYTHON_PREFIX}}/bin/pip install --no-cache-dir -r "{}"'.format(
+                            nb_requirements_file
+                        ),
+                    )
+                )
+
+        # install Pipfile.lock or fallback to installing Pipfile
+        pipenv = "${KERNEL_PYTHON_PREFIX}/bin/pipenv"
+        python = "${KERNEL_PYTHON_PREFIX}/bin/python"
+        pipfile = self.binder_path("Pipfile")
+        pipfile_lock = self.binder_path("Pipfile.lock")
+        working_directory = self.binder_dir or "."
+        assemble_scripts.append(("${NB_USER}", "pip install pipenv"))
+        # if Pipfile.lock isn't found, Pipfile is used to create one
+        if not os.path.exists(pipfile_lock):
+            assemble_scripts.append(
+                (
+                    "${NB_USER}",
+                    "(cd {} && {} lock --python {})".format(
```

Comment:
What is the working directory set to that we have to `cd`? I thought we'd be in the directory that contains the repo's content and that this is also where we wanted to be. If that isn't true maybe having a comment to explainn would help people from the future.

Reply:
Meant to be resolved by: c9fd0a9

---

The working directory we `cd` into is the repo directory or the `binder` / `.binder` directory.

If there is a reference to a `setup.py` file to be installed etc as there can be in `Pipfile`'s, then they will often reference either `.` or `..`. This has been verified to function with tests, and I've reasoned that this makes the most sense, as compared to making it so that you would need to write `.` within the Pipfile's residing in a binder directory and meaning the parent directory in reality.

### An example of a `Pipfile` outside a binder dir

```
[[source]]
url = "https://pypi.python.org/simple"
verify_ssl = true
name = "pypi"

[packages]
there = "*"
dummy = {path=".", editable=true}
```

### An example of a `Pipfile` in a binder dir

```
[[source]]
url = "https://pypi.python.org/simple"
verify_ssl = true
name = "pypi"

[packages]
there = "*"
dummy = {path="..", editable=true}
```

### The code we are discussing
```python
        # install Pipfile.lock or fallback to installing Pipfile
        pipenv = "${KERNEL_PYTHON_PREFIX}/bin/pipenv"
        python = "${KERNEL_PYTHON_PREFIX}/bin/python"
        pipfile = self.binder_path("Pipfile")
        pipfile_lock = self.binder_path("Pipfile.lock")
        working_directory = self.binder_dir or "."
        assemble_scripts.append(("${NB_USER}", "pip install pipenv"))
        # if Pipfile.lock isn't found, Pipfile is used to create one
        if not os.path.exists(pipfile_lock):
            assemble_scripts.append(
                (
                    "${NB_USER}",
                    "(cd {} && {} lock --python {})".format(
                        working_directory, pipenv, python
                    ),
                )
            )
        # install Pipfile.lock
        assemble_scripts.append(
            (
                "${NB_USER}",
                "(cd {} && {} install --ignore-pipfile --deploy --system --dev --python {})".format(
                    working_directory, pipenv, python
                ),
            )
        )
```